### DISCUSSION TITLE
Union type with scalar types and Data class not possible?

### DISCUSSION THREAD

**Discussion by ragulka** [spatie/laravel-data#743] [Q&A]
It seems that union types that include a `Data` class will always try to normalize the other type into `Data`. The following throws an error:

```
Could not create a `App\Data\TaxData` object from value with type `integer`, no normalizer was found.
```

```php
class OrderItem extends Data
{
    public function __construct(
        // either the tax rate object, or it's ID
        public int|TaxData $tax, 
        // ... other properties
    ) {}
}
```

I would expect that if `int` is an accepted type, and if the passed value is an `int`, this package would not try to cast it into the `TaxData` object.

I understand that I can use a custom cast, or even load the model and cast it to `TaxData` in the constructor, but given that there may be 100s of `OrderItem`s for a single order, it may mean that the same `Tax` models is loaded 100s of times from the DB. Instead, I'd like to skip loading the `Tax` models when receiving input from frontend, and only use `TaxData` when sending the data _to_ the frontend.

Any thoughts?

**Comment by rubenvanassche** [comment]
Technically we don't support union types, in versions before v4 we even let data throw an exception when an union type was used, that restriction was removed because I think in future versions we want to support this. But it is quite complicated since we need to rethink the casting, transforming, validation process.

The problem lies here:

https://github.com/spatie/laravel-data/blob/7eda411f0869162d96ef1981995ce68ca1ef95aa/src/DataPipes/CastPropertiesDataPipe.php#L84-L93

Since we allow data objects to be constructed from anything: arrays, strings, int's it is quite difficult to decide what should happen in this case. A MoneyData object could be totally valid to be constructed from a single int. We could check if a magic constructor exists for the property value but that would probably cost a lot of time.

What we could do, is trying to create the data object and if it fails just use the value. I've implemented this but it will not work with collectables since that's a whole other story. (https://github.com/spatie/laravel-data/commit/6075b4387371d3fa7db1afdf6712b63439a22573)
